File: backend/app/schemas/rewards.py
from decimal import Decimal
from pydantic import BaseModel, Field, ConfigDict
from typing import Optional, List, Dict, Any
from datetime import datetime

class CouponResponse(BaseModel):
    code: str
    description: str
    discount_type: str
    discount_value: Decimal
    min_cart_amount: Decimal
    max_discount_amount: Optional[Decimal] = None
    is_active: bool
    expires_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)

class VoucherResponse(BaseModel):
    id: str
    code: str
    title: str
    description: str
    discount_type: str
    discount_value: Decimal
    min_cart_amount: Decimal
    max_discount_amount: Optional[Decimal] = None
    expires_at: datetime
    status: str = "AVAILABLE" # "AVAILABLE", "USED", "EXPIRED"

    model_config = ConfigDict(from_attributes=True)

class CartPricingRequest(BaseModel):
    session_id: str
    coupon_code: Optional[str] = None
    voucher_code: Optional[str] = None
    use_coins: bool = False
    coins_to_redeem: Optional[int] = None

    model_config = ConfigDict(from_attributes=True)

class CartPricingResponse(BaseModel):
    subtotal: Decimal
    coupon_code: Optional[str] = None
    coupon_discount: Decimal = Decimal("0.00")
    voucher_code: Optional[str] = None
    voucher_discount: Decimal = Decimal("0.00")
    coins_used: int = 0
    coin_discount: Decimal = Decimal("0.00")
    delivery_charges: Decimal = Decimal("0.00")
    taxes: Decimal = Decimal("0.00")
    total: Decimal
    currency: str = "INR"
    points_to_earn: int = 0
    available_coin_balance: int = 0
    max_coins_redeemable: int = 0
    coin_value_inr: Decimal = Decimal("0.00")

    model_config = ConfigDict(from_attributes=True)

class CoinLedgerEntry(BaseModel):
    id: str
    amount: int # +credit, -debit
    transaction_type: str
    reference_id: Optional[str] = None
    description: str
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)

class RewardPointsLedgerEntry(BaseModel):
    id: str
    points: int # +credit, -debit
    transaction_type: str
    reference_id: Optional[str] = None
    description: str
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)

class CustomerRewardsSummary(BaseModel):
    coin_balance: int
    estimated_coin_value_inr: Decimal
    points_balance: int
    conversion_rate_description: str = "10 Apex Coins = ₹1.00"
    earning_rule_description: str = "1 Apex Point earned per ₹100 paid"
    available_vouchers: List[VoucherResponse] = []
    coin_history: List[CoinLedgerEntry] = []
    points_history: List[RewardPointsLedgerEntry] = []

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/lib/types/rewards.ts
export interface CouponData {
  code: string;
  description: string;
  discount_type: 'FIXED' | 'PERCENTAGE';
  discount_value: number;
  min_cart_amount: number;
  max_discount_amount?: number;
  is_active: boolean;
  expires_at?: string;
}

export interface VoucherData {
  id: string;
  code: string;
  title: string;
  description: string;
  discount_type: 'FIXED' | 'PERCENTAGE';
  discount_value: number;
  min_cart_amount: number;
  max_discount_amount?: number;
  expires_at: string;
  status: 'AVAILABLE' | 'USED' | 'EXPIRED';
}

export interface CartPricingBreakdown {
  subtotal: number;
  coupon_code?: string | null;
  coupon_discount: number;
  voucher_code?: string | null;
  voucher_discount: number;
  coins_used: number;
  coin_discount: number;
  delivery_charges: number;
  taxes: number;
  total: number;
  currency: string;
  points_to_earn: number;
  available_coin_balance: number;
  max_coins_redeemable: number;
  coin_value_inr: number;
}

export interface CoinLedgerItem {
  id: string;
  amount: number;
  transaction_type: string;
  reference_id?: string;
  description: string;
  created_at: string;
}

export interface RewardPointsLedgerItem {
  id: string;
  points: number;
  transaction_type: string;
  reference_id?: string;
  description: string;
  created_at: string;
}

export interface CustomerRewardsData {
  coin_balance: number;
  estimated_coin_value_inr: number;
  points_balance: number;
  conversion_rate_description: string;
  earning_rule_description: string;
  available_vouchers: VoucherData[];
  coin_history: CoinLedgerItem[];
  points_history: RewardPointsLedgerItem[];
}
